src/blocks/scratch3_pen.js

```diff
@@ -0,0 +1,231 @@
+ * @returns {PenState} the mutable pen state associated with that target. This will be created if necessary.
+ * @private
+ */
+Scratch3PenBlocks.prototype._getPenState = function (target) {
+    var penState = target.getCustomState(stateKey);
+    if (!penState) {
+        penState = JSON.parse(JSON.stringify(defaultPenState));
+        target.setCustomState(stateKey, penState);
+    }
+    return penState;
+};
+
+/**
+ * Handle a target which has moved. This only fires when the pen is down.
+ * @param {RenderedTarget} target - the target which has moved.
+ * @param {number} oldX - the previous X position.
+ * @param {number} oldY - the previous Y position.
+ * @private
+ */
+Scratch3PenBlocks.prototype._onTargetMoved = function (target, oldX, oldY) {
+    var penState = this._getPenState(target);
+    var penSkinId = this._getPenLayerID();
+    this.runtime.renderer.penLine(penSkinId, penState.penAttributes, oldX, oldY, target.x, target.y);
+    this.runtime.requestRedraw();
+};
+
+/**
+ * Update the cached RGB color from the hue & shade values in the provided PenState object.
+ * @param {PenState} penState - the pen state to update.
+ * @private
+ */
+Scratch3PenBlocks.prototype._updatePenColor = function (penState) {
+    var rgb = Color.hsvToRgb({h: penState.hue * 180 / 100, s: 1, v: 1});
+    var shade = (penState.shade > 100) ? 200 - penState.shade : penState.shade;
+    if (shade < 50) {
+        rgb = Color.mixRgb(Color.RGB_BLACK, rgb, (10 + shade) / 60);
+    } else {
+        rgb = Color.mixRgb(rgb, Color.RGB_WHITE, (shade - 50) / 60);
+    }
+    penState.penAttributes.color4f[0] = rgb.r / 255.0;
+    penState.penAttributes.color4f[1] = rgb.g / 255.0;
+    penState.penAttributes.color4f[2] = rgb.b / 255.0;
+};
+
+/**
+ * Wrap a pen hue or shade values to the range [0,200).
+ * @param {number} value - the pen hue or shade value to the proper range.
+ * @returns {number} the wrapped value.
+ * @private
+ */
+Scratch3PenBlocks.prototype._wrapHueOrShade = function (value) {
+    value = value % 200;
+    if (value < 0) value += 200;
+    return value;
+};
+
+/**
+ * Retrieve the block primitives implemented by this package.
+ * @return {Object.<string, Function>} Mapping of opcode to Function.
+ */
+Scratch3PenBlocks.prototype.getPrimitives = function () {
+    return {
+        pen_clear: this.clear,
+        pen_stamp: this.stamp,
+        pen_pendown: this.penDown,
+        pen_penup: this.penUp,
+        pen_setpencolortocolor: this.setPenColorToColor,
+        pen_changepencolorby: this.changePenHueBy,
+        pen_setpencolortonum: this.setPenHueToNumber,
+        pen_changepenshadeby: this.changePenShadeBy,
+        pen_setpenshadeto: this.setPenShadeToNumber,
+        pen_changepensizeby: this.changePenSizeBy,
+        pen_setpensizeto: this.setPenSizeTo
+    };
+};
+
+Scratch3PenBlocks.prototype.clear = function () {
+    var penSkinId = this._getPenLayerID();
+    this.runtime.renderer.penClear(penSkinId);
```

Comment:
I think we should always check to make sure the renderer exists. E.g.:
https://github.com/LLK/scratch-vm/blob/develop/src/sprites/rendered-target.js#L56